posts/2017-05-02-io-monad-for-cats.md

```diff
@@ -0,0 +1,404 @@
+
+## Enough Talk…
+
+What does this look like in practice?  Well, ideally, as convenient as possible!  Let's look at our println example:
+
+```scala
+def putStrLn(line: String): IO[Unit] =
+  IO { println(line) }
+
+f(putStrLn("hi!"), putStrLn("hi!"))
+
+// is equivalent to
+
+val x = putStrLn("hi!")
+f(x, x)
+```
+
+Great!  We can write Haskell fanfic in Scala.  😛
+
+The notable element here is the use of the `IO.apply` constructor to wrap the `println` effect in a *pure* `IO` value.  This pattern can be applied to any side-effect.  You can think of this sort of like an FFI that converts impure code (like `println`) into pure code (like `putStrLn`).  The goal of this API was to be as simple and straightforward as possible.  If you have a curly brace block of impure side-effecting code, you can wrap it in a composable and pure abstraction by just adding two characters: `IO`.  You can wrap arbitrarily large or small blocks of code, potentially involving complex allocations, JNI calls, resource semantics, etc; but it is generally considered a best practice to wrap side-effects into the smallest composable units that make sense and do all of your sequentialization using `flatMap` and `for`-comprehensions.
+
+For example, here's a program that performs some simple user interaction in the shell:
+
+```tut:silent
+import cats.effect.IO
+
+val program = for {
+  _ <- IO { println("Welcome to Scala!  What's your name?") }
+  name <- IO { Console.readLine }
+  _ <- IO { println(s"Well hello, $name!") }
+} yield ()
+```
+
+We could have just as easily written this program in the following way:
+
+```tut:silent
+val program = IO {
+  println("Welcome to Scala!  What's your name?")
+  val name = Console.readLine
+  println(s"Well hello, $name!")
+}
+```
+
+But this gives us less flexibility for composition.  Remember that even though `program` is a pure and referentially transparent value, its *definition* is not.  Anything inside the `IO {` … `}` block is not referentially transparent, and so should be treated with extreme care and suspicion.  The less of our program we have inside these blocks, the better!
+
```

Comment:
Just to hammer this home I would show what it means in terms of substitution. Something like

> … *definition* is not, which is to say, `IO(expr)` is **not** the same as `{ val x = expr; IO(x) }`. Anything …